Instance details page and event data docs (#848)
replicatedhq/replicated-docs: docs/vendor/monitoring-instance-details.md

# Monitoring Application Instances

This topic describes using the event data and key performance indicators (KPIs) in the Replicated vendor portal to monitor the health and performance of application instances installed in your customers' environments.

## About Instance Reporting

> Each time an active application instance installed in an online environment checks for updates, the Replicated app manager sends event data to the vendor portal with details about the application and the cluster where the application is installed.

Reviewer: ```suggestion
Each time an active application instance installed in an online environment checks for updates, the Replicated app manager sends a small amount of instance metadata to help help the update server build an appropriate list of available new versions for that instance. During this update check process, if any properties like app version or app status have changed,  an event is generated on the server side to record the state change.
```

The vendor portal uses this event data to provide insights about the health, status, and performance of the active application instances associated with each customer license. The vendor portal displays these insights on an **Instance details** dashboard. 

> You can use the event data and insights on the **Instance details** page to more quickly troubleshoot issues with active instances, helping to reduce support burden. The **Instance details** page also provides business value metrics for each instance, such as the rate of successful upgrades, total uptime, and the time that it takes your customers to successfully deploy an instance of your application.

Reviewer: ```suggestion
You can use the event data and insights on the **Instance details** page to more quickly troubleshoot issues with active instances, helping to reduce support burden. It becomes easy to understand answers to questions like:

1. Has the application experienced any downtime recently? How long has it been down?
1. Have any cluster or infrastructure changes occurred recently? Have nodes been lost or added? Has the underlying Kubernetes version changed?
1. Were any upgrades attempted recently? Which of them succeeded? Which version is the app currently running?

 
   The **Instance details** page also provides business value metrics for each instance, such as the rate of successful upgrades, total uptime, and the time that it takes your customers to successfully deploy an instance of your application.
```

* The **Instances details** page is available only for application instances installed in online environments. Data for instances installed in air gapped environments is not available.
* Inactive instances are not included in instance data. An instance is considered inactive if **ADD INACTIVE DESCRIPTION**
* The vendor portal receives updated instance data only when the application instance checks for updates. By default, instances automatically check for updates every four hours. However, if users edit the frequency of automatic update checks, then the rate at which instance data is updated changes accordingly. Additionally, if users disable automatic update checks, then the rate at which instance data is updated depends on the frequency that users manually check for updates. 
* To view event data and insights related to instance uptime on the **Instances details** page, you must configure status informers for your application in the Application custom resource. For more information about how to configure status informers, see [Displaying Application Status](admin-console-display-app-status).
* Data in the **Instance Activity** stream is not cached. **NEED MORE INFO**

> ## About the Instance Details Page {#about}

Reviewer: I never knew you could alias markdown headers like this! That's awesome 🙂

> Data for each active instance of your application is displayed in a dashboard on the **Instance details** page in the vendor portal. To access the **Instance details** page, go to **Customers** and click the **Customer reporting** button for the customer that you want to view. From the **Reporting** page for the selected customer, click the **View details** button for the desired application instance. 

Reviewer: do you think this could also use a small screenshot or diagram?

Reviewer: Maybe something like:

<img width="1166" alt="Screen Shot 2023-01-04 at 9 39 34 AM" src="https://user-images.githubusercontent.com/3730605/210592297-3a02fdf3-ccc5-4a05-baf1-cab2c7097dac.png">

Reviewer: or even just: 
<img width="813" alt="Screen Shot 2023-01-04 at 9 40 50 AM" src="https://user-images.githubusercontent.com/3730605/210592518-8882bbf6-ad88-4473-9ca7-f0cdec845afe.png">

Author: Yeah, definitely. Probably not super clear what the button looks like just from the text

The following screenshot shows an example of the **Instance details** page:

**UPDATE SCREENSHOT**

![Instance details full page](/images/instance-details.png)

[View a larger version of this image](/images/instance-details.png)

As shown in the screenshot above, the **Instance details** page includes the following sections:

* **Current State**: Displays information about the state of the instance, such as the current application version. See [Current State](#current-state) below.
* **Install Insights**: Displays KPIs related to health, performance, and adoption. See [Install Insights](#install-insights) below. 
* **Install Information**: Displays information about the cluster where the instance is installed, such as the version of Kubernetes running on the cluster and details about completed preflight checks. See [Install Information](#install-information) below.
* **Instance Uptime**: Displays granular details about instance uptime over time. See [Instance Uptime](#instance-uptime) below.

> * **Instance Activity**: Streams all event data from the instance. See [Instance Activity](#instance-activity) below.

Reviewer: This is great, and I like the summary section with links down to each header.

### Current State

The **Current State** section displays event data about the state of the instance at the time of the most recent update check.

**SCREENSHOT**

As shown in the screenshot above, the  **Current State** section includes the following fields:

> * **App status**: 

Reviewer: ```suggestion
* **App status**: The status computed based on an application's configured [Status Informers](/vendor/admin-console-display-app-status)
```

> * **App version**: 

Reviewer: ```suggestion
* **App version**: The version label of the currently running release, as set during [Release Promotion](/vendor/releases-creating-releases). If no version is set, the [release sequence](/vendor/releases-about#release-sequence-mechanics) will be used.
```

The vendor portal displays the following metrics that measure Time to Install:

* **Time to Install (License)**: The time between when you create the customer license in the vendor portal, and when the application instance `appStatus` field reports a `ready` state in the customer environment. Time to Install (License) represents the time that it takes for a customer to successfully deploy your application after you intend to distribute the application to the customer. Replicated uses the timestamp of when you create the customer license in the vendor portal to represent your intent to distribute the application to the customer because creating the license file for the customer is generally the final step before you share the installation materials with the customer.

   The following diagram demonstrates how the vendor portal computes the Time to Install (License) metric:
   
   ** **INSERT DIAGRAM** **

>    As shown in the diagram above, Time to Install (License) includes several activities that are involved in deploying the application, including the customer receiving the necessary materials and documentation, downloading the assets, provisioning the required hardware, networking, external systems, completing the preflight checks, and finally installing, configuring, and deploying the application.

Reviewer: This is awesome, good to have this extra clarity

* **Time to Install (Instance)**: The time between when the vendor portal records the first event for the application instance in the customer environment, and when the instance `appStatus` field reports a `ready` state.

   The following diagram demonstrates the activities included and excluded in Time to Install (Instance):

   ** **INSERT DIAGRAM** **

>    As shown in the diagram above, Time to Install (Instance) is the length of time that it takes for the application to reach a Ready state after the customer makes a successful deployment attempt in their environment. A deployment attempt is considered to be successful when the application instance reports its first event. Time to Install (Instance) does not include any deployment attempts that were unsuccessful. For example, Time to Live (Instance) does not include any time spent by the customer discarding servers and attempting to deploy the instance again on a new server.

Reviewer: This needs some work but I don't have specific pointers -- I'm not on board with this definition of "successful deployment attempt"

Reviewer: A deployment attempt is **started** when the application reports its first event. It's not **successful** until the first ready status

Author: Ah, okay. I think the language you use there is good. So the idea is more that it's tracking the time between when 
a deployment is "started" (first event) and when there's a "successful" deployment (Ready status).

And I guess there could many attempts within that time span? As in, I start a deployment and the vendor portal records an event for an instance, but then I have to try a few more times before I can actually get a Ready status

#### Upgrades

​​The ease with which customers can upgrade their application instance is often an important factor in determining the quality of a delivered software package. If the process of upgrading is simple and straightforward, customers may be more likely to upgrade their software, which can help to ensure that they are using the most up-to-date version. On the other hand, if the process is complicated, error-prone, or time-consuming, customers may be less likely to upgrade, which can result in them using outdated or vulnerable versions of the software.

The vendor portal displays the following metrics to measure the ease of upgrading your application:

> * **Number of Upgrades**: The vendor portal computes the total number of upgrades using the `appVersion` and `appStatus` events.

Reviewer: **Number of Upgrades** and **Upgrade Success Rate** will probably be out of scope for MVP, need to hide temporarily

   TBD

* **Upgrade Success Rate**: The vendor portal calculates the upgrade success rate by dividing the number of successful upgrades by the total number of upgrades completed.
  
  The vendor portal considers an upgrade successful if _all_ of the following are true:     
    * There are no errors on the deployment of the new version. Depending on if the customer upgrades the instance using the app manager or the helm CLI, then the deployment command is either `kubectl apply` or `helm upgrade`.
    * The instance enters a Ready status in 15 minutes or less after the `kubectl apply` or `helm upgrade` command completes.
    * The instance does _not_ enter a Missing or Unavailable status in the first 15 minutes after the instance is no longer in an Updating status.

#### Uptime

> The vendor portal computes the total uptime for the instance as the fraction of time that an application spends in a Ready, Updating, or Degraded state. High uptime indicates that the application is reliable and able to handle the demands of the customer environment, while low uptime might indicate that the software is prone to errors or failures. By measuring the total uptime of customer instances, you can better understand the performance of your application.

Reviewer: ```suggestion
The vendor portal computes the total uptime for the instance as the fraction of time that an application spends in a Ready, Updating, or Degraded state. High uptime indicates that the application is reliable and able to handle the demands of the customer environment, while low uptime might indicate that the application is prone to errors or failures. By measuring the total uptime of customer instances, you can better understand the performance of your application.
```

The vendor portal uses the status of the Kubernetes resources that you provide in the `statusInformers` field of the Application custom resource to determine the status of the application. For more information about how to configure status informers for your application, see [Displaying Application Status](admin-console-display-app-status).

The following table lists the application statuses that indicate whether the application is up or down:

<table>
  <tr>
    <th>Up</th>
    <td>Ready, Updating, Degraded</td>
  </tr>
  <tr>
    <th>Down</th>

>     <td>Missing, Unavailable, Inactive</td>

Reviewer: ```suggestion
    <td>Missing, Unavailable</td>
```

:::

The following shows an example of an Instance Uptime graph:

**INSERT SCREENSHOT**

You can hover over the bars in the Instance Uptime graph to view more detail about the percent of time that the instance was in each state during the given time period.

### Instance Activity

The **Instance Activity** section displays event data for the instance. The event data stream is updated each time an instance checks for updates. When an instance checks for updates, it reports details about the current app manager version, Kubernetes version, application version, and application status. 

By default, instances automatically check for updates every four hours. However, if users edit the frequency of automatic update checks, then the rate at which instance data is updated changes accordingly. Additionally, if users disable automatic update checks, then the rate at which instance data is updated depends on the frequency that users manually check for updates.

The following shows an example of the **Instance Activity** data stream:

**SCREENSHOT**

You can filter the **Instance Activity** stream by the following event categories:

* TBD
* TBD
* TBD

> For more information about the event data displayed in the **Instance Activity** stream, see [Event Data](monitoring-event-data).

Reviewer: Probably needs some information about inactive instances when we ship that part, but that can come later, not a blocker to publish this.

## About the Instance Details Page {#about-page}

The vendor portal provides insights about the health, status, and performance of the active application instances associated with each customer license on the **Instance details** page. You can use the insights on the **Instance details** page to more quickly troubleshoot issues with your customers' active instances, helping to reduce support burden. 

For example, you can use the **Instance details** page to track the following events for each instance:

* Recent performance degradation or downtime
* Length of instance downtime
* Recent changes to the cluster or infrastructure
* Changes in the number of nodes, such as nodes lost or added
* Changes in the cluster's Kubernetes version
* Changes in the application version that the instance is running

> To access the **Instance details** page, go to **Customers** and click the **Customer reporting** button for the customer that you want to view:

Reviewer 2: Here and throughout both topics, I recommend changing "click the **nameOfButton** button to just "click **nameOfButton". We have that captured in our Style Guide: https://wiki.replicated.io/doc/grammar-and-formatting-28QRVfWVgb#h-ui-buttons

But it's just a nit.

Author: I went back and forth on this one...I ended up leaving the word "button" in this case since there is no text on the screen. As in, it didn't seem the same as saying "Click the Save button", since in that case, the button would actually have the word "Save" on it.

It didn't feel totally accurate to say "click **View details**", when there is nothing on the page that obviously says "View details". (It felt like we'd need to call out that it's a button that has the label "View details" when you hover over it).

What do you think?

Reviewer 2: Ah, ok, it's that darned icon. Agreed, let's leave it with "button" included. We might want to capture that exception in the Style Guide.

  :::

### Install Information

The **Install Information** section displays the following details about the cluster infrastructure where the application is installed:

* The Kubernetes distribution for the cluster. For example, GKE or EKS.
* The version of Kubernetes running in the cluster.
* The Replicated app manager (KOTS) version installed in the cluster.
* For **First Seen**, the timestamp of the first event that the vendor portal generated for the instance. For more information about how the vendor portal generates events, see [About Events](monitoring-event-data#about-events) in _Event Data_.
* If detected, the cloud provider and region where the cluster is running. For example, `GCP: us-central1`.

In addition to the details listed above, the **Install Information** section also displays the following for clusters provisioned by the Replicated Kubernetes installer:
* Node operating systems
* Node operating systems versions
* Total number of cluster nodes

> * Number of cluster nodes in a ready state

Reviewer 2: Capitalization: "ready" --> "Ready"

import Checkins from "../partials/instance-monitoring/_appCheckins.mdx"

> # Monitoring Application Instances

Reviewer: Per comment below, want a word other than `Monitoring`. As of yesterday, the working title for this body of features is `Instance Insights` - could consider using that

```suggestion
# Application Instance Insights
```

> This topic describes using the insights in the Replicated vendor portal to monitor the health and performance of application instances installed in your customers' environments.

Reviewer: I think monitoring is the wrong word here. Monitoring implies regular or passive observation, perhaps even with Alerting baked in, whereas the key "job to be done" with this page is "quickly get up to speed on what's happening with an instance", almost always in response to an internal status inquiry or a customer support request. Added some more detail in the doc: https://docs.google.com/document/d/1GsyL0s2LPNrYsk-Ih1t_TYmyMBQaXaJCd8be2RR2xK4/edit#heading=h.c3gaka77i3fr

This suggestion might need wordsmithing

```suggestion
This topic describes using the insights in the Replicated vendor portal to quickly understand the recent events and performance of application instances installed in your customers' environments. While there are many use cases for this data, this is generally most valuable when diagnosing issues with customer instances.
```

Author: Got it--good clarification. I'll update the title and do a search for the term "monitoring" throughout to edit

## About Instance Reporting {#about-reporting}

For active application instances installed in online customer environments, the Replicated app manager sends a small amount of instance data to the vendor portal when any of the following occur:

<Checkins/>

The instance data that the app manager sends includes properties such as the current application version and application status. The primary purpose of this instance data is to compile the list of new versions that are available to the given instance for upgrade.

The vendor portal also uses this data to display insights about the active instances of your application. For more information about the instance data insights displayed in the vendor portal, see [About the Instance Details Page](#about-page) below.

## Requirements and Limitations

Instance data has the following requirements and limitations:

> * You must configure status informers for your application in the Application custom resource to view instance data about the application status or uptime. For more information about how to configure status informers, see [Displaying Application Status](admin-console-display-app-status).

Reviewer: ```suggestion
* You must configure one or more status informers for your application in the Application custom resource to populate instance data about the application status or uptime. For more information about how to configure status informers, see [Displaying Application Status](admin-console-display-app-status).
```

* Instance data is available only for application instances installed in online environments. Data for instances installed in air gapped environments is not available.

> * Instance data is available only for active application instances. An instance is considered inactive if **ADD INACTIVE DESCRIPTION**

Reviewer: We pushed inactive stuff past MVP, will document what we do now:

```suggestion
* Instance data is available only for active application instances. If an instance is decommissioned, stops checking for updates, or otherwise stops reporting, it will continue to be displayed in the most-recently seen state. This may cause confusion if an instance was ready and then its VM was destroyed before it reported again. You can use the "Last Check-in" field to understand if an instance's data may have become stale.
```

* The rate at which data is updated on the **Instances details** page varies depending on how often the vendor portal receives instance data from the app manager. The vendor portal receives instance data when any of the following occur:

  <Checkins/>

> * The timestamp of events displayed on the **Instances details** page is the timestamp when the vendor portal received the instance data from the app manager. The timestamp of events does not necessarily reflect the timestamp of when the event occurred.

Reviewer: ```suggestion
* The timestamp of events displayed on the **Instances details** page is the timestamp when the Replicated instance-facing API received the instance data from the app manager. The timestamp of events does not necessarily reflect the timestamp of when the event occurred.
```

Author: Is this the Vendor API? docs.replicated.com/reference/vendor-api-using

> * The app manager stores the `kurlNodeCountTotal` and `kurlNodeCountReady` data fields in a cache for five minutes. If the app manager sends instance data to the vendor portal within the five minute window, then the reported data for `kurlNodeCountTotal` and `kurlNodeCountReady` reflects the data in the cache. This means that events displayed on the **Instances details** page for the `kurlNodeCountTotal` and `kurlNodeCountReady` fields can show values that differ from the current values of these fields.

Reviewer: ```suggestion
* The app manager stores the `kURL Total Cluster Nodes` and `kURL Ready Nodes`  data fields in a cache for five minutes. If the app manager sends instance data to the vendor portal within the five minute window, then the reported data for `kURL Total Nodes` and `kURL Ready Nodes` reflects the data in the cache. This means that events displayed on the **Instances details** page for the `kURL Total Nodes` and `kURL Ready Nodes` fields can show values that differ from the current values of these fields.
```

## About the Instance Details Page {#about-page}

The vendor portal provides insights about the health, status, and performance of the active application instances associated with each customer license on the **Instance details** page. You can use the insights on the **Instance details** page to more quickly troubleshoot issues with your customers' active instances, helping to reduce support burden. 

For example, you can use the **Instance details** page to track the following events for each instance:

* Recent performance degradation or downtime
* Length of instance downtime
* Recent changes to the cluster or infrastructure
* Changes in the number of nodes, such as nodes lost or added

> * Changes in the Kubernetes version in the cluster

Reviewer: ```suggestion
* Changes in the cluster's Kubernetes version
```
